Assertion failure with default interface method

## Steps to Reproduce

1. Compile the attached repro (or use precompiled exe): `csc /unsafe /debug dimbug.cs`
2. Run it: `mono dimbug.exe`

[dimbug.zip](https://github.com/mono/mono/files/3876044/dimbug.zip)

<!--
You may drag & drop the attachment (repro code/solution, screenshot, etc.) onto the issue.
-->

### Current Behavior

```
$ mono dimbug.exe 
* Assertion: should not be reached at mini-trampolines.c:590


=================================================================
	Native Crash Reporting
=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.
=================================================================

=================================================================
	Native stacktrace:
=================================================================
	0x10d4768db - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : mono_dump_native_crash_info
	0x10d40ef15 - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : mono_handle_native_crash
	0x10d475e9b - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : sigabrt_signal_handler
	0x7fff6ba2db1d - /usr/lib/system/libsystem_platform.dylib : _sigtramp
	0x0 - Unknown
	0x7fff6b903a1c - /usr/lib/system/libsystem_c.dylib : abort
	0x10d672dc7 - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : monoeg_assert_abort
	0x10d653cff - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : mono_log_write_logfile
	0x10d67314e - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : monoeg_g_logv_nofree
	0x10d6732cf - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : monoeg_assertion_message
	0x10d411f9b - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : common_call_trampoline
	0x10d412a2d - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : mono_vcall_trampoline
	0x10d89b513 - Unknown
	0x10d9671a9 - Unknown
	0x10d967396 - Unknown
	0x10d36c5ab - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : mono_jit_runtime_invoke
	0x10d578dd7 - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : mono_runtime_invoke_checked
	0x10d58145e - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : do_exec_main_checked
	0x10d3cca07 - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : mono_jit_exec
	0x10d3cfc1d - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : mono_main
	0x10d35b4e8 - /Library/Frameworks/Mono.framework/Versions/Current/Commands/mono : main
	0x7fff6b82c2e5 - /usr/lib/system/libdyld.dylib : start
	0x2 - Unknown

=================================================================
	Telemetry Dumper:
=================================================================
Pkilling 0x700003795000 from 0x11905edc0
Entering thread summarizer pause from 0x11905edc0
Finished thread summarizer pause from 0x11905edc0.

Waiting for dumping threads to resume

=================================================================
	External Debugger Dump:
=================================================================
(lldb) command source -s 0 '/tmp/mono-gdb-commands.36305'
Executing commands in '/tmp/mono-gdb-commands.36305'.
(lldb) process attach --pid 36305
error: attach failed: Error 1

=================================================================
	Basic Fault Address Reporting
=================================================================
Memory around native instruction pointer (0x7fff6b97b49a):0x7fff6b97b48a  ff ff c3 90 90 90 b8 48 01 00 02 49 89 ca 0f 05  .......H...I....
0x7fff6b97b49a  73 08 48 89 c7 e9 13 9f ff ff c3 55 48 89 e5 41  s.H........UH..A
0x7fff6b97b4aa  57 41 56 41 55 41 54 53 48 81 ec 88 00 00 00 4d  WAVAUATSH......M
0x7fff6b97b4ba  89 cf 49 89 f5 48 8d 05 82 75 ef 29 48 8b 00 48  ..I..H...u.)H..H

=================================================================
	Managed Stacktrace:
=================================================================
	  at <unknown> <0xffffffff>
	  at MainClass:Main <0x00128>
	  at System.Object:runtime_invoke_void <0x000a5>
=================================================================
Abort trap: 6
```

### Expected Behavior

Should not abort.

## On which platforms did you notice this

- [x] macOS
- [ ] Linux
- [ ] Windows

**Version Used**:

```
$ mono --version
Mono JIT compiler version 6.8.0.52 (2019-10/7a71151252e Wed Nov 20 12:10:34 EST 2019)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	Interpreter:   yes
	LLVM:          yes(610)
	Suspend:       hybrid
	GC:            sgen (concurrent by default)
```

## Stacktrace

See above.

<!--
You can join us at https://gitter.im/mono/mono to discuss your reported issue
-->

Interpreter throws a managed exception:
```console
$ ./mono/mini/mono-sgen --interp dimbug.exe

Unhandled Exception:
System.InvalidOperationException: Could not execute the method because the containing type is not fully instantiated.
  at MainClass.Main () <0x7fdf8704d2a8 + 0x00046> in <ca30c6c9482448538d4437ece4d73e3d>:0
[ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidOperationException: Could not execute the method because the containing type is not fully instantiated.
  at MainClass.Main () <0x7fdf8704d2a8 + 0x00046> in <ca30c6c9482448538d4437ece4d73e3d>:0
```

--- a/mono/metadata/class-init.c
+++ b/mono/metadata/class-init.c
@@ -3027,11 +3027,11 @@ mono_class_setup_vtable_general (MonoClass *klass, MonoMethod **overrides, int o
 		for (int i = 0; i < iface_onum; i++) {
 			MonoMethod *decl = iface_overrides [i*2];
 			MonoMethod *override = iface_overrides [i*2 + 1];
-			if (mono_class_is_gtd (override->klass))
-				override = mono_class_inflate_generic_method_full_checked (override, ic, mono_class_get_context (ic), error);
 			if (decl->is_inflated) {
 				override = mono_class_inflate_generic_method_checked (override, mono_method_get_context (decl), error);
 				mono_error_assert_ok (error);
+			} else if (mono_class_is_gtd (override->klass)) {
+				override = mono_class_inflate_generic_method_full_checked (override, ic, mono_class_get_context (ic), error);
 			}
 			if (!apply_override (klass, ic, vtable, decl, override, &override_map, &override_class_map, &conflict_map))
 				goto fail;

--- a/mono/metadata/class-init.c
+++ b/mono/metadata/class-init.c
@@ -3027,6 +3027,8 @@ mono_class_setup_vtable_general (MonoClass *klass, MonoMethod **overrides, int o
 		for (int i = 0; i < iface_onum; i++) {
 			MonoMethod *decl = iface_overrides [i*2];
 			MonoMethod *override = iface_overrides [i*2 + 1];
+			if (mono_class_is_gtd (override->klass))
+				override = mono_class_inflate_generic_method_full_checked (override, ic, mono_class_get_context (ic), error);
 			if (decl->is_inflated) {
 				override = mono_class_inflate_generic_method_checked (override, mono_method_get_context (decl), error);
 				mono_error_assert_ok (error);

--- a/mono/tests/Makefile.am
+++ b/mono/tests/Makefile.am
@@ -384,6 +384,7 @@ TESTS_CS_SRC=		\
 	reflection.cs		\
 	interface.cs		\
 	interface-2.cs		\
+	dim-generic.cs		\
 	iface.cs		\
 	iface2.cs		\
 	iface3.cs		\

--- a/mono/tests/dim-generic.cs
+++ b/mono/tests/dim-generic.cs
@@ -1,0 +1,34 @@
+using System;
+
+
+interface IBaseThingy
+{
+	int Foo ();
+}
+
+interface INativeThingy<T> : IBaseThingy
+{
+	int IBaseThingy.Foo () {
+		return 0;
+        }
+}
+
+class NativeThingy : INativeThingy<string>
+{
+}
+
+public class Test
+{
+	public static int test_0_dim_override()
+	{
+		var thingy = new NativeThingy ();
+		var ithingy = (IBaseThingy)thingy;
+		int i = ithingy.Foo ();
+		return i;
+    }
+
+    public static int Main (string[] args) {
+		return TestDriver.RunTests (typeof (Test), args);
+	}
+
+}

